### DISCUSSION TITLE
How to use RedfishPkg

### DISCUSSION THREAD

**Discussion by Timboooh** [tianocore/edk2#4077] [Q&A]
Hey,
I'm currently developing my first UEFI application, and it needs to communicate to a Redfish API.
I'm compiling the RedfishPkg using:
```
TARGET = DEBUG
TARGET_ARCH = X64
TOOL_CHAIN_TAG = GCC5
ACTIVE_PLATFORM = RedfishPkg/RedfishPkg.dsc
```
After successfully compiling this produces some efi files in ./Build/RedfishPkg/RELEASE_GCC5/X64
But when I try to load the confighandler in the UEFI shell, provided by OVMF, I get this error:
```
FS0:\> load RedfishConfigHandlerDriver.efi
Image 'FS0:\RedfishConfigHandlerDriver.efi' error in StartImage: Not Found
```
My qemu logs contains the following messages:
```
FSOpen: Open '\RedfishConfigHandlerDriver.efi' Success
[Security] 3rd party image[0] can be loaded after EndOfDxe: PciRoot(0x0)/Pci(0x1,0x1)/Ata(Primary,Master,0x0)/HD(1,MBR,0xBE1AFDFA,0x3F,0xFBFC1)/\RedfishConfigHandlerDriver.efi.
InstallProtocolInterface: 5B1B31A1-9562-11D2-8E3F-00A0C969723B 7DF95040
Loading driver at 0x0007DF74000 EntryPoint=0x0007DF75A09 RedfishConfigHandlerDriver.efi
InstallProtocolInterface: BC62157E-3E33-4FEC-9920-2D3B36D750DF 7DF7A218
ProtectUefiImageCommon - 0x7DF95040
  - 0x000000007DF74000 - 0x0000000000002A00
Error: Image at 0007DF74000 start failed: Not Found
```

Because I'm really inexperienced in EDK2 I have no clue how to debug this issue, and how to correctly load and use the RedfishPkg. Any pointers would be greatly appreciated!

Thanks,
Timo

**Comment by nicklela** [comment]
RedfishConfigHandlerDriver will normally be dispatched by DXE dispatcher when all dependencies are satisfied. If we like to load it manually, we have to handle dependency on ourselves. 

The dependency for RedfishConfigHandlerDriver can be found in INF file: 
https://github.com/tianocore/edk2/blob/c1dd400a13d1a5e862809c55f6760ba3a944a609/RedfishPkg/RedfishConfigHandler/RedfishConfigHandlerDriver.inf#L58-L60

It requires Redfish Credential Protocol to be ready and functional. This is the way that Redfish driver can get the credential  and talk to BMC.  You can check here for the architecture design of Redfish driver stack: https://github.com/tianocore/edk2/blob/master/RedfishPkg/Readme.md#edk2-redfish-implementation-diagrams
